xarray/core/duck_array_ops.py

```diff
@@ -386,6 +386,38 @@ def count(data, axis=None):
     return xp.sum(xp.logical_not(isnull(data)), axis=axis)
 
 
+def nunique(data, axis=None, skipna=True):
+    """Count the number of unique values in this array along the given axis or axes"""
+    xp = get_array_namespace(data)
+
+    if axis is None:
+        axis = list(range(data.ndim))
+    elif isinstance(axis, (int, tuple)):
+        axis = [axis] if isinstance(axis, int) else list(axis)
+
+    # If axis empty, return unchanged data.
+    if not axis:
+        return data
+
+    # Move axes to be aggregated to the end and stack
+    shape = data.shape
+    new_order = [i for i in range(len(shape)) if i not in axis] + axis
+    new_shape = [s for i, s in enumerate(shape) if i not in axis] + [-1]
+    stacked = xp.reshape(xp.transpose(data, new_order), new_shape)
+
+    # Check if data has type object; if so use pd.factorize for unique integers
```

Comment:
This is a good start but we'd prefer a vectorized approach. 

I'd start with the first answer here: https://stackoverflow.com/questions/46893369/count-unique-elements-along-an-axis-of-a-numpy-array
but replace the `np.diff` with `not_equal(a[..., :-1], a[..., 1:])` and sum that along the axis. You'll have to handle the case of NaNs not comparing equal, presumably by also summing `duck_array_ops.isnull(a)` along the same axis, and subtracting it away.